# Overview

This notebook assumes you have already collected and scored MOC tweets. It creates a dataset for use in R to analyze the patterns of polarization over time. You will do some parsing on an AWS server and some locally before ultimately making a CSV file that you can open and analyze in R.

LH Note: on my computer, Git stuff and data live in different places, so you'll see notes about moving files or changing directories. I haven't figured out a good way to keep files in both places or to mirror or sync or whatever. So, for now, paths are hard-coded or there's a note about where to find a file.

# Get Data

2016 election data is on an AWS server under ```/data/purpletag```. 

To login: 

```ssh -i ~/.ssh/carolgrrr.pem ubuntu@purpletag.casmlab.org```

The data is large (> 4GB), so best to run Juypter notebooks to parse on the server. Then CSV files can be used locally.

You can run a notebook on the server and use your local browser with these two commands:

* ```ssh -L 8080:localhost:8888 -i ~/.ssh/carolgrrr.pem ubuntu@purpletag.casmlab.org```
* ```nohup jupyter notebook --no-browser > log.txt 2>&1 &```

Then access ```http://localhost:8080``` in your browser.

## On server: Parsing from ```scores``` files to CSV

This section assumes you have already run purpletag's ```collect``` and ```score``` functions and gotten the Twitter data that you want in JSON format and parsed that data into score files.

In [ ]:
# Based on https://stackoverflow.com/questions/26415906/read-multiple-txt-files-into-pandas-dataframe-with-filename-as-column-header
import pandas as pd
import os
import glob

# Step 1: get a list of all score files in target directory
my_dir = "/data/purpletag/scores/"
filelist = []
filesList = []
os.chdir( my_dir )

# Step 2: Build up list of files:
for files in glob.glob("*.1.moc.scores"): # using 1-day purpletag MOC scores
    fileName, fileExtension = os.path.splitext(files)
    filesList.append(files) #filename with extension

In [ ]:
# Step 3: Build up DataFrame:
# Based on https://stackoverflow.com/questions/35717706/python-how-to-turn-a-dictionary-of-dataframes-into-one-big-dataframe-with-colum
d = {} # dictionary to hold multiple dfs

for filename in filesList:
    df1 = pd.read_csv(filename, header=None, sep=' ', index_col=0)
    d[filename[:-13]] = df1
    
df = pd.concat(d, axis=1)
df.columns = df.columns.droplevel(-1) 

df.to_pickle('/data/purpletag/mocs_by_date.pkl')

Move the file from the AWS server to local if you want to work locally:

```scp -i ~/.ssh/carolgrrr.pem ubuntu@purpletag.casmlab.org:/data/purpletag/mocs_by_date.pkl ~/Documents/git/casmlab/purpletag/2016-election-study/data-files/```

## Locally: Prepping for stats

We now have a pickled dataframe of the form handleXdate. We need to keep data only from Labor Day to Election Day and get weekly averages.

In [16]:
import pandas as pd

df = pd.read_pickle('data-files/mocs_by_date.pkl')
df.head()

2015-11-10  2015-11-11  2015-11-12  2015-11-13  2015-11-14  \
austinscottga08         NaN     2.40933         NaN         NaN         NaN   
benniegthompson         NaN         NaN         NaN         NaN         NaN   
bettymccollum04         NaN         NaN         NaN         NaN         NaN   
billpascrell            NaN         NaN   -1.168750   -0.916501   -0.972477   
boblatta                NaN         NaN    0.100723         NaN         NaN   

                 2015-11-15  2015-11-16  2015-11-17  2015-11-18  2015-11-19  \
austinscottga08         NaN         NaN         NaN         NaN         NaN   
benniegthompson         NaN         NaN     3.85585         NaN         NaN   
bettymccollum04         NaN         NaN         NaN         NaN         NaN   
billpascrell            NaN         NaN         NaN         NaN         NaN   
boblatta            1.70286         NaN     1.56897         NaN   -0.031978   

                    ...      2016-10-30  2016-10-31  2016-11-01  2016-11-02  \
austinscottga08     ...             NaN         NaN         NaN         NaN   
benniegthompson     ...             NaN         NaN         NaN         NaN   
bettymccollum04     ...             NaN         NaN     -90.838    -61.3435   
billpascrell        ...             NaN         NaN         NaN    -17.9542   
boblatta            ...             NaN         NaN         NaN         NaN   

                 2016-11-03  2016-11-04  2016-11-05  2016-11-06  2016-11-07  \
austinscottga08         NaN         NaN         NaN         NaN         NaN   
benniegthompson         NaN         NaN         NaN         NaN         NaN   
bettymccollum04     -40.482         NaN    -33.4513         NaN         NaN   
billpascrell            NaN         NaN         NaN         NaN         NaN   
boblatta                NaN         NaN         NaN         NaN         NaN   

                 2016-11-08  
austinscottga08         NaN  
benniegthompson         NaN  
bettymccollum04         NaN  
billpascrell            NaN  
boblatta                NaN  

[5 rows x 365 columns]

In [17]:
week1_dates = ['2016-09-06','2016-09-07','2016-09-08','2016-09-09','2016-09-10','2016-09-11','2016-09-12']
week2_dates = ['2016-09-13','2016-09-14','2016-09-15','2016-09-16','2016-09-17','2016-09-18','2016-09-19']
week3_dates = ['2016-09-20','2016-09-21','2016-09-22','2016-09-23','2016-09-24','2016-09-25','2016-09-26']
week4_dates = ['2016-09-27','2016-09-28','2016-09-29','2016-09-30','2016-10-01','2016-10-02','2016-10-03']
week5_dates = ['2016-10-04','2016-10-05','2016-10-06','2016-10-07','2016-10-08','2016-10-09','2016-10-10']
week6_dates = ['2016-10-11','2016-10-12','2016-10-13','2016-10-14','2016-10-15','2016-10-16','2016-10-17']
week7_dates = ['2016-10-18','2016-10-19','2016-10-20','2016-10-21','2016-10-22','2016-10-23','2016-10-24']
week8_dates = ['2016-10-25','2016-10-26','2016-10-27','2016-10-28','2016-10-29','2016-10-30','2016-10-31']
week9_dates = ['2016-11-01','2016-11-02','2016-11-03','2016-11-04','2016-11-05','2016-11-06','2016-11-07']

df['week1'] = df[week1_dates].mean(axis=1)
df['week2'] = df[week2_dates].mean(axis=1)
df['week3'] = df[week3_dates].mean(axis=1)
df['week4'] = df[week4_dates].mean(axis=1)
df['week5'] = df[week5_dates].mean(axis=1)
df['week6'] = df[week6_dates].mean(axis=1)
df['week7'] = df[week7_dates].mean(axis=1)
df['week8'] = df[week8_dates].mean(axis=1)
df['week9'] = df[week9_dates].mean(axis=1)

df.head()

2015-11-10  2015-11-11  2015-11-12  2015-11-13  2015-11-14  \
austinscottga08         NaN     2.40933         NaN         NaN         NaN   
benniegthompson         NaN         NaN         NaN         NaN         NaN   
bettymccollum04         NaN         NaN         NaN         NaN         NaN   
billpascrell            NaN         NaN   -1.168750   -0.916501   -0.972477   
boblatta                NaN         NaN    0.100723         NaN         NaN   

                 2015-11-15  2015-11-16  2015-11-17  2015-11-18  2015-11-19  \
austinscottga08         NaN         NaN         NaN         NaN         NaN   
benniegthompson         NaN         NaN     3.85585         NaN         NaN   
bettymccollum04         NaN         NaN         NaN         NaN         NaN   
billpascrell            NaN         NaN         NaN         NaN         NaN   
boblatta            1.70286         NaN     1.56897         NaN   -0.031978   

                  ...     2016-11-08      week1       week2      week3  \
austinscottga08   ...            NaN        NaN    4.932035        NaN   
benniegthompson   ...            NaN        NaN   -1.211380        NaN   
bettymccollum04   ...            NaN -56.669016 -119.884760 -67.172900   
billpascrell      ...            NaN  -2.871430   -2.623810  -2.353348   
boblatta          ...            NaN   1.379970    9.673527        NaN   

                     week4      week5      week6     week7      week8    week9  
austinscottga08   1.064290   3.675700   6.042600  1.561610        NaN      NaN  
benniegthompson        NaN        NaN  -1.823010       NaN        NaN      NaN  
bettymccollum04 -59.540265 -27.995798 -33.955595 -4.879673 -33.080633 -56.5287  
billpascrell     -2.337897        NaN  -1.026550       NaN  -1.333330 -17.9542  
boblatta               NaN        NaN   0.974138       NaN        NaN      NaN  

[5 rows x 374 columns]

In [18]:
weekly_df = df[['week1','week2','week3','week4','week5','week6','week7','week8','week9']]
weekly_df

week1       week2      week3      week4      week5  \
austinscottga08        NaN    4.932035        NaN   1.064290   3.675700   
benniegthompson        NaN   -1.211380        NaN        NaN        NaN   
bettymccollum04 -56.669016 -119.884760 -67.172900 -59.540265 -27.995798   
billpascrell     -2.871430   -2.623810  -2.353348  -2.337897        NaN   
boblatta          1.379970    9.673527        NaN        NaN        NaN   
bradsherman            NaN   -1.941180        NaN  -5.381950        NaN   
call_me_dutch    -1.312583   -4.765343  -3.129013  -1.573270  -2.967365   
candicemiller     1.678195         NaN   0.029657  -0.969076        NaN   
cathymcmorris     7.926215    3.141282   8.873570  13.830632   1.000000   
cbrangel        -39.434357  -61.711175 -41.164824 -37.333904 -19.490345   
chelliepingree   -0.478424   -1.141239   0.001002  -3.231895  -0.645854   
chrismurphyct   -10.588038  -42.525738  -4.712589  -4.560957  -6.453698   
chrisvanhollen  -54.966055  -99.947165 -10.479500        NaN -13.279100   
chuckgrassley          NaN    0.688525   1.228070   1.193205   0.906780   
clyburn          -0.957143 -125.927880 -19.132180 -49.990600        NaN   
conawaytx11            NaN         NaN        NaN        NaN        NaN   
congbillposey     0.967105         NaN        NaN        NaN        NaN   
congculberson     2.049355   16.449665        NaN   5.953135   0.906780   
conghuelskamp     3.189410    5.334424   9.784920  12.915598  10.034006   
congmikesimpson   1.694790    0.757014   1.076198   1.064290   2.427180   
congpalazzo       2.779320         NaN  -1.052340  16.833300        NaN   
congressmangt     5.996984   12.966840   8.609298   6.305415   5.557500   
congressmanhice   5.782418   22.958759   3.558067   2.783482   2.340362   
congressmanruiz  -0.619814         NaN        NaN  -2.495040  -0.966387   
corybooker      -45.880645   -4.211167 -10.528431  -9.191768  -4.854678   
cynthialummis     0.941390         NaN  -0.000332        NaN   1.699210   
darrellissa            NaN    6.615326   1.780650        NaN        NaN   
daveloebsack      1.379970    1.141780  -1.222368  -4.729487        NaN   
davereichert      1.749340    9.632034   4.175330   2.660523   4.601563   
desjarlaistn04    2.008113    6.360580   2.063235        NaN   1.675186   
...                    ...         ...        ...        ...        ...   
sentomcotton      9.259097   18.984000   5.861856   2.213970   2.213375   
sentoomey         0.450846    1.340321        NaN   1.730260        NaN   
senwarren       -16.472357  -20.003685        NaN  -9.442620        NaN   
senwhitehouse   -26.853336  -27.274468  -5.569098  -6.160863  -2.246048   
speakerryan      10.366340   37.134494  20.404184  24.263986  18.337240   
staceyplaskett  -74.466400  -64.339300 -21.400700  -3.613390 -14.520628   
stevedaines       0.614514   15.955496   3.194646   2.988327   1.992373   
stevekingia       0.884615         NaN        NaN   1.058930        NaN   
steveknight25    -0.033462         NaN        NaN   0.129049        NaN   
stevescalise     12.777685   17.993253  11.207510  11.448580  12.882075   
susanwbrooks     -1.859935   20.762442   3.977370   2.799142   0.906780   
teammoulton     -10.223317  -39.223645  -9.080928  -6.886165 -14.323355   
tgowdysc          1.379970   24.362623   5.717880   6.417177        NaN   
toddrokita        3.874347    6.490850   2.564010  17.983067   3.175610   
tomcoleok04      -0.927027    1.490305   0.890323        NaN        NaN   
tomrooney              NaN    0.916522        NaN        NaN        NaN   
tulsipress       -5.130997   -9.110062  -4.423757 -10.482710  -4.345324   
txrandy14         2.778870    2.538835   2.618932   6.070819        NaN   
usrepdavidjolly  -8.071250   -7.510378        NaN   1.077513   1.843037   
usrepgarypalmer  -0.602145    4.222761   1.906190  11.043693   1.240658   
usrepkcastor    -44.156545  -44.457523 -45.454540  -6.229427 -30.346366   
usrepkeating     -2.224717    0.067697  -0.905

In [19]:
import pandas as pd
import yaml

# get the data from Govtrack
with open('/Users/libbyh/Dropbox/CASM/SMCE/Shared Social Media and Civic Engagement/Data/purpletag/legislators-social-media.yaml', 'r') as f:
    df_social = pd.io.json.json_normalize(yaml.load(f))

with open('/Users/libbyh/Dropbox/CASM/SMCE/Shared Social Media and Civic Engagement/Data/purpletag/legislators-current.yaml', 'r') as f:
    df_current = pd.io.json.json_normalize(yaml.load(f))

print(len(weekly_df))
# merge everything into one data frame with one row per MOC
df_meta = pd.merge(df_current, df_social, on="id.govtrack")
df_meta["handle"] = df_meta["social.twitter"].str.lower()
weekly_df["handle"] = weekly_df.index.str.lower()

print(len(df_meta))

df_merged = pd.merge(df_meta, weekly_df, left_on="handle", right_index=True)

print(len(df_merged))

#cols_to_keep = ['id.govtrack','social.twitter','name.official_full','bio.gender','terms','week1','week2','week3','week4','week5','week6','week7','week8','week9']

df_merged = df_merged[['id.govtrack','social.twitter','name.official_full','bio.gender','terms','week1','week2','week3','week4','week5','week6','week7','week8','week9']]


511
529
444


/Users/libbyh/Documents/VirtualEnvs/purpletag/lib/python3.6/site-packages/ipykernel/__main__.py:15: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy


Not sure why we have only 444 matches, but it's better than 12.

In [20]:
df1 = pd.concat([df_merged.drop(['terms'], axis=1), df_merged['terms'].apply(pd.Series)], axis=1)
df2 = pd.concat([df1.drop([0], axis=1), df1[0].apply(pd.Series)], axis=1)

keep_df = df2[['id.govtrack','social.twitter','name.official_full','bio.gender','type','party','week1','week2','week3','week4','week5','week6','week7','week8','week9']]
keep_df

id.govtrack   social.twitter      name.official_full bio.gender type  \
0         400050  SenSherrodBrown           Sherrod Brown          M  rep   
1         300018  SenatorCantwell          Maria Cantwell          F  rep   
2         400064    SenatorCardin      Benjamin L. Cardin          M  rep   
3         300019    SenatorCarper        Thomas R. Carper          M  rep   
4         412246      SenBobCasey    Robert P. Casey, Jr.          M  sen   
5         412248     SenBobCorker              Bob Corker          M  sen   
6         300043     SenFeinstein        Dianne Feinstein          F  sen   
7         300052    SenOrrinHatch          Orrin G. Hatch          M  sen   
9         412243  McCaskillOffice        Claire McCaskill          F  sen   
10        400272  SenatorMenendez         Robert Menendez          M  rep   
11        300078    SenBillNelson             Bill Nelson          M  rep   
13        300093      SenStabenow         Debbie Stabenow          F  rep   
14        412244    SenatorTester              Jon Tester          M  sen   
15        412247    SenWhitehouse      Sheldon Whitehouse          M  sen   
16        412251  SenJohnBarrasso           John Barrasso          M  sen   
17        400432    SenatorWicker         Roger F. Wicker          M  rep   
18        300002     SenAlexander         Lamar Alexander          M  sen   
19        300023   SenThadCochran            Thad Cochran          M  rep   
20        300025   SenatorCollins        Susan M. Collins          F  sen   
21        300027       JohnCornyn             John Cornyn          M  sen   
22        300038    SenatorDurbin       Richard J. Durbin          M  rep   
23        300041      SenatorEnzi         Michael B. Enzi          M  sen   
24        300047       GrahamBlog          Lindsey Graham          M  rep   
25        300055      InhofePress         James M. Inhofe          M  rep   
26        300072   McConnellPress         Mitch McConnell          M  sen   
27        412325   SenJeffMerkley            Jeff Merkley          M  sen   
28        300081      SenJackReed               Jack Reed          M  rep   
29        412322     SenatorRisch          James E. Risch          M  sen   
30        300083    SenPatRoberts             Pat Roberts          M  rep   
31        412323   SenatorShaheen          Jeanne Shaheen          F  sen   
..           ...              ...                     ...        ...  ...   
438       412641   RepDavidRouzer            David Rouzer          M  rep   
439       412643  RepTomMacArthur        Thomas MacArthur          M  rep   
440       412644        RepBonnie   Bonnie Watson Coleman          F  rep   
441       412646     RepLeeZeldin           Lee M. Zeldin          M  rep   
442       412647  RepKathleenRice        Kathleen M. Rice          F  rep   
443       412648      RepStefanik       Elise M. Stefanik          F  rep   
444       412649     RepJohnKatko              John Katko          M  rep   
445       412650       RepRussell           Steve Russell          M  rep   
446       412651  RepRyanCostello        Ryan A. Costello          M  rep   
448       412653     RepRatcliffe          John Ratcliffe          M  rep   
449       412654    hurdonthehill               Will Hurd          M  rep   
450       412655    RepBrianBabin             Brian Babin          M  rep   
451       412656       repmialove             Mia B. Love          F  rep   
452       412657      repdonbeyer    Donald S. Beyer, Jr.          M  rep   
453       412658      RepComstock        Barbara Comstock          F  rep   
454       412659   staceyplaskett      Stacey E. Plaskett          F  rep   
455       412660      RepNewhouse            Dan Newhouse          M  rep   
456       412661      RepGrothman          Glenn Grothman          M  rep   
457       412662    RepAlexMooney     Alexander X. Mooney          M  rep   
458       412663   RepEvanJenkins         Evan H. Jenkins          M  rep   
4

In [21]:
# melt it so each row is a person x week
df_long = pd.melt(keep_df, id_vars=['id.govtrack','social.twitter','name.official_full','bio.gender','party','type'],
                value_vars=['week1','week2','week3','week4','week5','week6','week7','week8','week9'],
                var_name='week', value_name='avg_score')
df_long['week'] = df_long['week'].str[-1:]

df_long.rename(columns = {'type':'chamber', 'social.twitter': 'handle', 'name.official_full': 'name', 'bio.gender': 'gender'}, inplace = True)

df_long.head()

id.govtrack           handle                  name gender     party  \
0       400050  SenSherrodBrown         Sherrod Brown      M  Democrat   
1       300018  SenatorCantwell        Maria Cantwell      F  Democrat   
2       400064    SenatorCardin    Benjamin L. Cardin      M  Democrat   
3       300019    SenatorCarper      Thomas R. Carper      M  Democrat   
4       412246      SenBobCasey  Robert P. Casey, Jr.      M  Democrat   

  chamber week  avg_score  
0     rep    1  -0.873145  
1     rep    1 -42.572318  
2     rep    1  -8.692816  
3     rep    1 -38.936383  
4     sen    1 -19.997507

In [22]:
# make sure we have just two parties
df_long.party.unique()

array(['Democrat', 'Republican'], dtype=object)

In [23]:
# get an absolute value of the polar score
df_long['abs'] = df_long['avg_score'].abs()

In [24]:
df_long.to_csv('data-files/weekly_averages_long.csv')

# Now move to R for analysis

Run ```~/Documents/git/casmlab/purpletag/2016_election.R```

That R script sends its output to ```2016_election_results.txt```

In [180]:
results = open('data-files/2016_election_results.txt', 'r')
print(results.read())


> # for pretty regression tables
> # http://stackoverflow.com/questions/30195718/stargazer-save-to-file-dont-show-in-console
> mod_stargazer <- functi .... [TRUNCATED] 

> df <- read.csv('weekly_averages_long.csv', header = TRUE, sep = ",", quote = "\"",
+                dec = ".", fill = TRUE, comment.char = "")

> summary(df)
       X           id.govtrack                 handle                 name      gender  
 Min.   :   0.0   Min.   :300002   AustinScottGA08:   9   Adam B. Schiff:   9   F: 792  
 1st Qu.: 998.8   1st Qu.:400326   BennieGThompson:   9   Adam Kinzinger:   9   M:3204  
 Median :1997.5   Median :412292   BettyMcCollum04:   9   Adam Smith    :   9           
 Mean   :1997.5   Mean   :401868   BillPascrell   :   9   Adrian Smith  :   9           
 3rd Qu.:2996.2   3rd Qu.:412533   BobLatta       :   9   Al Franken    :   9           
 Max.   :3995.0   Max.   :412674   BradSherman    :   9   Al Green      :   9           
                                   (Other)    

Based on the outlier-excluded linear mixed-effects models, it makes sense to remove RepThompson. The pattern stays the same even with RepThompson in the set though: negative effect for republican and week, positive effect for their interaction. ```lmm5``` is the model-of-best-fit. 

## Changing the way we score hashtags

What if we score tags for the 63-day period and then score MOCS?

Run the following (on the server) to get new scores:

* purpletag parse -t 63 -d 200
* purpletag score
* ~~purpletag score --counts --score-mocs~~

That first command took a week because the code starts with today and works backwards 200 days, one day at a time. Each day takes over an hour. See Issue #18 about options for changing this behavior.

With the new tag measures, can start the process over. Start at "On Server: Parsing scores to CSV" with a new file name.

Score MOCs didn't work quite right, so doing it here.

In [12]:
import io

def get_tag_scores(score_file):
    scores = {}
    with io.open(score_file, encoding='utf8') as f:
        for line in f:
            (key, val) = line.split()
            scores[key] = val
    return scores

def score_mocs(score_file, counts):
    """ Output a file with user scores. """
    scores = get_tag_scores(score_file)

    mocs = {}
    # get the MOC's tags
    for line in inf:
        moc_score = 0
        tags = {}
        parts = line.strip().split()
        handle = parts[0]
        tags = parse_tags(parts[1:], args)

        # calculate a MOC score
        for tag, count in tags.iteritems():
            if not counts:
                try:
                    moc_score += float(scores[tag])
                except KeyError:
                    # print "Can't find key for", tag
                    continue
            else:
                moc_score += float(scores[tag]) * count
        mocs[handle] = moc_score

In [13]:


# handle2party = twitter_handle_to_party()
# print handle2party.items()[0]

score_file = "data-files/2016-11-09.63.scores"
print('scoring MOCs with', score_file)
score_mocs(score_file, 'true')

scoring MOCs with data-files/2016-11-09.63.scores


NameError: name 'inf' is not defined